MAINT: numpy cleanup version bumps: fixes issue #20458

peppedilillo opened the pull request:
The minimum numpy version in `scipy/scipy/__init__.py` has been aligned to the minimum requirement of `pyproject.toml`. A numpy version check has been removed from test `test_moments` in `scipy/stats/tests/test_multivariate.py` because the check looked for a numpy version earlier than the minimum. This pull request does not fully solve #20458. I will come back at it after discussion.

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Partially solves issue #20458 

#### What does this implement/fix?
This aligns a few numpy versions. Does not align the pytest version.

#### Additional information
I'm open to working on fixing the pytest versions too, as reported in #20458. However, it is not clear to me what is the exact pytest version we should report. Shall we find it ourselves or you already know? Tried grepping for it but didn't find anything useful.

peppedilillo committed:
MAINT: numpy cleanup version bumps: partially fixes issue #20458

The minimum numpy version in `scipy/scipy/__init__.py` has been aligned to the minimum requirement of `pyproject.toml`.
A numpy version check has been removed from test `test_moments` in `scipy/stats/tests/test_multivariate.py` because the check looked for a numpy version earlier than the minimum.
This pull request does not fully solve #20458. I will come back at it after discussion.

tylerjereddy commented on scipy/__init__.py:
```diff
@@ -66,7 +66,7 @@
 from scipy._lib import _pep440
 # In maintenance branch, change to np_maxversion N+3 if numpy is at N
-np_minversion = '1.22.4'
+np_minversion = '2.0.0rc1'
```
This is not correct, it would ban imports of NumPy prior to the 2.x series, which we still allow.

peppedilillo replied:
I realize this is a serious error, sorry! Also I found PR#20459 just after committing.. I have pushed a fix to the problem you report, and a few changes which should resolve the pytest version guard problem. Sorry for the duplicate and, again, for the unfortunate error!

tylerjereddy reviewed (commented):
This PR is basically a duplicate of https://github.com/scipy/scipy/pull/20459 with an additional problem at the moment.

Also, the point in question, the `pytest` version guard, was already discussed over there at: https://github.com/scipy/scipy/pull/20459#pullrequestreview-2012055393

peppedilillo committed:
MAINT: numpy cleanup version bumps: fixes issue #20458

After checking discussion to PR #20459, found that the pytest guards could be removed.
So I did, and removed a couple imports and a function which would became unused too.
Also, this commit fixes an horrible error with the previous one, which would ban import prior to NumPy 2.0.0.
(sorry! and thanks @tylerjereddy)
To fix this I change the numpy minimum version of `scipy/__init__.py` to `1.23.5` as user JacobOgle did with PR #20459.
Thanks Jacob!

tylerjereddy commented:
The CI failures were unrelated, and the diff seems "ok" now, so I've squash-merged it. Note that I've added @JacobOgle as a co-author giving the two active PRs.

peppedilillo commented:
> The CI failures were unrelated, and the diff seems "ok" now, so I've squash-merged it. Note that I've added @JacobOgle as a co-author given the two active PRs.

Thank you. Great to have added @JacobOgle: i copypasted his numpy version! Thank you too Jakob! 
Bests and catch you soon hopefully. ~peppe
